src/routes/version_creation.rs

```diff
@@ -0,0 +1,239 @@
+use crate::models::mods::{
+    GameVersion, ModId, ModLoader, Version, VersionFile, VersionId, VersionType,
+};
+use crate::routes::mod_creation::{CreateError, UploadedFile};
+use actix_multipart::{Field, Multipart};
+use actix_web::web::Data;
+use actix_web::{post, HttpResponse};
+use futures::stream::StreamExt;
+use serde::{Deserialize, Serialize};
+use sqlx::postgres::PgPool;
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct InitialVersionData {
+    pub mod_id: Option<ModId>,
+    pub file_parts: Vec<String>,
+    pub version_number: String,
+    pub version_title: String,
+    pub version_body: String,
+    pub dependencies: Vec<VersionId>,
+    pub game_versions: Vec<GameVersion>,
+    pub release_channel: VersionType,
+    pub loaders: Vec<ModLoader>,
+}
+
+#[post("api/v1/version")]
+pub async fn version_create(
+    payload: Multipart,
+    client: Data<PgPool>,
+    file_host: Data<std::sync::Arc<dyn FileHost + Send + Sync>>,
+) -> Result<HttpResponse, CreateError> {
+    let mut transaction = client.begin().await?;
+    let mut uploaded_files = Vec::new();
+
+    let result = version_create_inner(
+        payload,
+        &mut transaction,
+        &***file_host,
+        &mut uploaded_files,
+    )
+    .await;
+
+    if result.is_err() {
+        let undo_result = super::mod_creation::undo_uploads(&***file_host, &uploaded_files).await;
+        let rollback_result = transaction.rollback().await;
+
+        if let Err(e) = undo_result {
+            return Err(e);
+        }
+        if let Err(e) = rollback_result {
+            return Err(e.into());
+        }
+    } else {
+        transaction.commit().await?;
+    }
+
+    result
+}
+
+async fn version_create_inner(
+    mut payload: Multipart,
+    transaction: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+    file_host: &dyn FileHost,
+    uploaded_files: &mut Vec<UploadedFile>,
+) -> Result<HttpResponse, CreateError> {
+    //TODO: Check if mod exists, check if version number is unique
+    let cdn_url = dotenv::var("CDN_URL")?;
+
+    let mut initial_version_data = None;
+    let mut version_builder = None;
+
+    while let Some(item) = payload.next().await {
+        let mut field: Field = item.map_err(CreateError::MultipartError)?;
+        let content_disposition = field.content_disposition().ok_or_else(|| {
+            CreateError::MissingValueError("Missing content disposition".to_string())
+        })?;
+        let name = content_disposition
+            .get_name()
+            .ok_or_else(|| CreateError::MissingValueError("Missing content name".to_string()))?;
+
+        if name == "data" {
+            let mut data = Vec::new();
+            while let Some(chunk) = field.next().await {
+                data.extend_from_slice(&chunk.map_err(CreateError::MultipartError)?);
+            }
+
+            let version_create_data: InitialVersionData = serde_json::from_slice(&data)?;
+            initial_version_data = Some(version_create_data.clone());
```

Comment:
This clone seems unnecessary.  You may be able to avoid the extra allocations by making `version_create_data` a reference to the value that has been moved into the option.
```rust
let version_create_data: InitialVersionData = serde_json::from_slice(&data)?;
initial_version_data = Some(version_create_data);
let version_create_data = initial_version_data.as_ref().unwrap();
```